src/rp2040/hardware_structs/include/hardware/structs/sio.h

```diff
@@ -11,51 +13,162 @@
 #include "hardware/regs/sio.h"
 #include "hardware/structs/interp.h"
 
+// reference to datasheet: https://datasheets.raspberrypi.org/rp2040/rp2040-datasheet.pdf#tab-registerlist_sio
+
+// BITMASK [BITRANGE]: FIELDNAME (RESETVALUE): DESCRIPTION
+// The REG macro is intended to help make the register navigable in your IDE (for example, using the "Go to Definition" feature)
+// _REG_(x) will link to the corresponding register in hardware/regs/sio.h.
+
 typedef struct {
+    _REG_(SIO_CPUID_OFFSET)
+    // Processor core identifier
     io_ro_32 cpuid;
+
+    _REG_(SIO_GPIO_IN_OFFSET)
+    // Input value for GPIO pins
+    // 0x3fffffff [0-29]  : GPIO_IN (0): Input value for GPIO0
     io_ro_32 gpio_in;
+
+    _REG_(SIO_GPIO_HI_IN_OFFSET)
+    // Input value for QSPI pins
+    // 0x0000003f [0-5]   : GPIO_HI_IN (0): Input value on QSPI IO in order 0
     io_ro_32 gpio_hi_in;
-    uint32_t _pad;
 
+    uint32_t _pad0;
+
+    _REG_(SIO_GPIO_OUT_OFFSET)
+    // GPIO output value
+    // 0x3fffffff [0-29]  : GPIO_OUT (0): Set output level (1/0 -> high/low) for GPIO0
     io_rw_32 gpio_out;
-    io_wo_32 gpio_set;
-    io_wo_32 gpio_clr;
-    io_wo_32 gpio_togl;
 
-    io_wo_32 gpio_oe;
-    io_wo_32 gpio_oe_set;
-    io_wo_32 gpio_oe_clr;
-    io_wo_32 gpio_oe_togl;
+    _REG_(SIO_GPIO_OUT_SET_OFFSET)
+    // GPIO output value set
+    // 0x3fffffff [0-29]  : GPIO_OUT_SET (0): Perform an atomic bit-set on GPIO_OUT, i
+    io_rw_32 gpio_set;
```

Comment:
I wonder why this was previously `io_wo_32 gpio_set;` when the datasheet says `GPIO_OUT_SET` is RW? Perhaps the datasheet "should" list the all atomic bit-manipulation GPIO registers in the SIO block (`GPIO_xxxx_SET`, `GPIO_xxxx_CLR` and `GPIO_xxxx_XOR`) as WO? :man_shrugging:

Reply:
I mean arguably they are write-only -- you get garbage when you read, and all-zeroes just happens to be the cheapest possible garbage to generate. I had a quick look at the hardware source and there is a comment saying `// write-only`. So I guess this is more of a philosophical question.

Reply:
Thanks @lurch